### PR TITLE
DAOS-9195 vos: Prevent multiple range discard on same object

### PR DESCRIPTION
**Pull request by jolivier23**
Master PR #7637 
Master PR #7463

### CODE REVIEW

**Comment by johannlombardi** [comment]
Could you please clarify whether the ask is to land this one or PR 7642?

**Comment by wangdi1** [comment]
> Could you please clarify whether the ask is to land this one or PR #7642?

please land this one.

**Review by wangdi1 — DISMISSED** [review]

**Comment by johannlombardi** [comment]
@NiuYawei @gnailzenh @liuxuezhao @Nasf-Fan  please prioritize this inspection on Monday. That's the last 2.0 rc1 blocker. Thanks in advance!

File: src/container/srv_target.c [hunk 1]
```diff
@@ -170,24 +170,19 @@ cont_aggregate_runnable(struct ds_cont_child *cont, struct sched_request *req,
...
+		D_DEBUG(DB_EPC, DF_CONT": skip aggregation during reintegration %d.\n",
+			DP_CONT(cont->sc_pool->spc_uuid, cont->sc_uuid),
+			pool->sp_reintegrating);
+		return false;
 	}
 
+	if (!cont->sc_agg_active)
+		D_DEBUG(DB_EPC, DF_CONT": resume aggregation after reintegration.\n",
+			DP_CONT(cont->sc_pool->spc_uuid, cont->sc_uuid));
+	cont->sc_agg_active = 1;
```
Comment (Nasf-Fan) [inline src/container/srv_target.c:184]: Can be inside above "if(!cont->sc_agg_active)" section.

File: src/object/srv_obj.c [hunk 2]
```diff
@@ -2340,7 +2340,8 @@ ds_obj_tgt_update_handler(crt_rpc_t *rpc)
 			  (orw->orw_bulks.ca_arrays != NULL ||
 			   orw->orw_bulks.ca_count != 0) ? true : false);
 	if (rc != 0)
-		D_CDEBUG(rc == -DER_INPROGRESS || rc == -DER_TX_RESTART, DB_IO, DLOG_ERR,
+		D_CDEBUG(rc == -DER_EXIST || rc == -DER_NONEXIST || rc == -DER_INPROGRESS ||
```
Comment (Nasf-Fan) [inline src/object/srv_obj.c:2343]: Can be more accurate to distinguish conditional insert or update, that can be done in other PR, such as https://github.com/daos-stack/daos/pull/7598

**Review by Nasf-Fan — DISMISSED** [review]

**Comment by johannlombardi** [comment]
hm, OSA offline reintegration & drain failed twice, even after retriggering it. Is that a regression?

**Comment by johannlombardi** [comment]
Looks like https://daosio.atlassian.net/browse/DAOS-9179 ?

**Comment by wangdi1** [comment]
>hm, OSA offline reintegration & drain failed twice, even after retriggering it. Is that a regression?

reintegration failed by this

2021-12-12 10:01:31,374 process          L0416 DEBUG| [stderr] ERROR (aiori-DFS.c:559): 0: 5: Failed to mount DFS namespace
2021-12-12 10:01:31,379 process          L0416 DEBUG| [stdout] ERROR: Failed to initialize DAOS DFS driver, (aiori-DFS.c:594)
2021-12-12 10:01:31,379 process          L0416 DEBUG| [stderr] application called MPI_Abort(MPI_COMM_WORLD, -1) - process 0

**Comment by wangdi1** [comment]
ah, exceed RF
12/12-10:01:31.37 wolf-129 DAOS[50805/50805/0] dfs  ERR  src/client/dfs/dfs.c:1187 open_sb() Failed to fetch SB info, DER_RF(-2031): 'Failures exceed RF'

**Comment by wangdi1** [comment]
Ah, it seems a test script issue.

The test kill rank4, then suppose to reintegrate the rank totally,  but somehow it only reintegrate 2 targets of rank4, which has 8 totally.

2021-12-12 10:01:04,060 process          L0604 INFO | Running '/usr/bin/dmg -o /etc/daos/daos_control.yml -d pool reintegrate A0638BA9-C1C2-42EA-9B10-818719526B3A --rank=4 --target-idx=0,2'


Then try to re-open the container, so got RF failure.

**Comment by wangdi1** [comment]
>Looks like https://daosio.atlassian.net/browse/DAOS-9179 ?

Yes, offline drain is due to 9179.

**Review by liuxuezhao — APPROVED** [review]

**Review by liuxuezhao — DISMISSED** [review]

File: src/container/srv_target.c [hunk 3]
```diff
@@ -170,24 +170,19 @@ cont_aggregate_runnable(struct ds_cont_child *cont, struct sched_request *req,
...
-				DP_CONT(cont->sc_pool->spc_uuid, cont->sc_uuid),
-				pool->sp_reintegrating);
-			return false;
-		}
-		if (!cont->sc_vos_agg_active)
-			D_DEBUG(DB_EPC, DF_CONT": resume aggregation after reintegration.\n",
-				DP_CONT(cont->sc_pool->spc_uuid, cont->sc_uuid));
-		cont->sc_vos_agg_active = 1;
+	if (pool->sp_reintegrating) {
+		cont->sc_agg_active = 0;
```
Comment (NiuYawei) [inline src/container/srv_target.c:174]: See the comments from PR-7637

**Review by NiuYawei — CHANGES_REQUESTED** [review]

File: src/container/srv_target.c (hunk 3 above)
Comment (wangdi1) [reply src/container/srv_target.c]: ah, you are right. Just updated the patch. please check.

File: src/container/srv_target.c [hunk 4]
```diff
@@ -564,6 +569,7 @@ cont_ec_agg_ult(void *arg)
 	D_DEBUG(DB_EPC, "start EC aggregation "DF_UUID"\n",
 		DP_UUID(cont->sc_uuid));
 
+	cont->sc_ec_agg_active = 1;
```
Comment (NiuYawei) [inline src/container/srv_target.c:572]: Why this extra set is required?

File: src/object/srv_obj_migrate.c [hunk 5]
```diff
@@ -2642,7 +2641,8 @@ destroy_existing_obj(struct migrate_pool_tls *tls, unsigned int tgt_idx,
 	}
 
 	/* Wait until container aggregation are stopped */
-	while (cont->sc_vos_agg_active && !tls->mpt_fini && !cont->sc_stopping) {
+	while (cont->sc_ec_agg_active && cont->sc_vos_agg_active &&
```
Comment (NiuYawei) [inline src/object/srv_obj_migrate.c:2644]: (sc_ec_agg_active || sc_vos_agg_active), right?

**Review by NiuYawei — CHANGES_REQUESTED** [review]

**Review by NiuYawei — APPROVED** [review]

**Review by Nasf-Fan — APPROVED** [review]

**Pull request merged**